monai/engines/trainer.py

```diff
@@ -145,3 +146,145 @@ def _iteration(self, engine: Engine, batchdata: Union[Dict, Sequence]) -> Dict[s
+        3. Generate m fakes from new random latent codes.
+        4. Update generator with these fakes using discriminator feedback.
+
+    Based on [Goodfellow 2014] https://arxiv.org/abs/1406.2661
+
+    Args:
+        device: an object representing the device on which to run.
+        max_epochs: the total epoch number for engine to run.
+        train_data_loader: Real data DataLoader to update discriminator.
+        g_network: generator (G) network architecture.
+        g_optimizer: G optimizer function.
+        g_loss_function: G loss function for optimizer.
+        d_network: discriminator (D) network architecture.
+        d_optimizer: D optimizer function.
+        d_loss_function: D loss function for optimizer.
+        g_inferer: inference method to execute G model forward on latent code.
+        d_inferer: inference method to execute D model forward.
+        d_train_steps: number of times to update D with real data minibatch.
+        latent_shape: size of G random input latent code.
+        make_latent: callback function to create batch of latent input for G inferer.
+        prepare_batch: callback function to prepare DataLoader output for D inferer.
+        iteration_update: the callable function for every iteration, expect to accept `engine`
+            and `batchdata` as input parameters. if not provided, use `self._iteration()` instead.
+        amp: whether to enable auto-mixed-precision training, reserved.
+        post_transform: execute additional transformation for the model output data.
+            Typically, several Tensor based transforms composed by `Compose`.
+        key_train_metric: compute metric when every iteration completed, and save average value to
+            engine.state.metrics when epoch completed. key_train_metric is the main metric to compare and save the
+            checkpoint into files.
+        additional_metrics: more Ignite metrics that also attach to Ignite Engine.
+        train_handlers: every handler is a set of Ignite Event-Handlers, must have `attach` function, like:
+            CheckpointHandler, StatsHandler, SegmentationSaver, etc.
+
+    """
+
+    def __init__(
+        self,
+        device: torch.device,
+        max_epochs: int,
+        train_data_loader: DataLoader,
+        g_network: torch.nn.Module,
+        g_optimizer: Optimizer,
+        g_loss_function: Callable,
+        d_network: torch.nn.Module,
+        d_optimizer: Optimizer,
+        d_loss_function: Callable,
+        g_inferer: Inferer = SimpleInferer(),
+        d_inferer: Inferer = SimpleInferer(),
+        d_train_steps: int = 1,
+        latent_shape: int = 64,
+        prepare_batch: Callable = default_prepare_batch,
+        make_latent: Callable = make_rand_latent_code,
+        iteration_update: Optional[Callable] = None,
+        amp: bool = True,
+        post_transform: Optional[Transform] = None,
+        key_train_metric: Optional[Dict[str, Metric]] = None,
```

Comment:
do we need to expose this `key_train_metric` option?

Reply:
I think for future implementations yes. Like StyleGAN or BigGAN, many computer vision GANs use distribution comparison calculations every n epoch's as a key training metric (FID, PPL, IS)

Also since some GAN networks collapse at scale, this key_train_metric could be used with early stopping Monai handlers.